lib/ex_hls/client/live/reader.ex

```diff
@@ -0,0 +1,383 @@
+    |> Enum.each(fn
+      %ExM3U8.Tags.Discontinuity{} ->
+        Logger.warning("""
+        [ExHLS.Client] Discontinuity tag found in the M3U8 media playlist. \
+        This may cause issues in further handling the stream.
+        """)
+
+      _other_tag ->
+        :ok
+    end)
+  end
+
+  defp download_chunks(state) do
+    {maybe_media_init, state} = get_media_inits_to_download(state)
+
+    download_from_seq_num =
+      next_segment_to_download_seq_num(state)
+
+    segments_to_download_with_seq_nums =
+      state.media_playlist.timeline
+      |> Enum.filter(&match?(%Segment{}, &1))
+      |> Enum.with_index(state.media_playlist.info.media_sequence)
+      |> Enum.filter(fn {_segment, seq_num} -> seq_num >= download_from_seq_num end)
+
+    {segments_to_download, seq_nums} = segments_to_download_with_seq_nums |> Enum.unzip()
+
+    state =
+      state
+      |> Map.update!(:max_downloaded_seq_num, fn
+        nil -> seq_nums |> Enum.max(&>=/2, fn -> nil end)
+        number -> [number | seq_nums] |> Enum.max()
+      end)
+
+    (maybe_media_init ++ segments_to_download)
+    |> Enum.reduce(state, &download_and_consume_segment/2)
+    |> maybe_end_stream()
+  end
+
+  defp get_media_inits_to_download(%{media_init_downloaded?: true} = state), do: {[], state}
+
+  defp get_media_inits_to_download(state) do
+    media_inits =
+      state.media_playlist.timeline
+      |> Enum.flat_map(fn
+        %MediaInit{} = media_init -> [media_init]
+        _other_tag -> []
+      end)
+
+    {media_inits, %{state | media_init_downloaded?: true}}
+  end
+
+  defp next_segment_to_download_seq_num(%{max_downloaded_seq_num: nil} = state) do
+    {segments_with_end_times, duration_sum} =
+      state.media_playlist.timeline
+      |> Enum.flat_map_reduce(0, fn
+        %Segment{} = segment, duration_acc ->
+          duration_acc = duration_acc + segment.duration
+          {[{segment, duration_acc}], duration_acc}
+
+        _other_tag, duration_acc ->
+          {[], duration_acc}
+      end)
+
+    start_time = duration_sum - 2 * state.media_playlist.info.target_duration
+
+    segment_index =
+      segments_with_end_times
+      |> Enum.find_index(fn {_segment, segment_end_time} ->
+        start_time <= segment_end_time
+      end)
+
+    segment_index + state.media_playlist.info.media_sequence
+  end
+
+  defp next_segment_to_download_seq_num(%{max_downloaded_seq_num: last_seq_num}) do
+    last_seq_num + 1
+  end
+
+  defp schedule_next_playlist_check(%{end_stream_executed?: true} = state) do
+    Logger.warning("[ExHLS.Client] Not scheduling next playlist check, end of playlist reached")
+    state
+  end
+
+  defp schedule_next_playlist_check(%{playlist_check_scheduled?: false} = state) do
+    last_segment =
+      state.media_playlist.timeline
+      |> Enum.reverse()
+      |> Enum.find(&match?(%Segment{}, &1))
+
+    assumed_duration_s =
+      case last_segment do
+        %Segment{duration: duration} -> duration
```

Comment:
I am hesitant to that, as with this implementation we would send GET playlist request every `1/4*last_segment_duration` for the whole duration of the "playlist window". If that last_segment was really short, we could be banned by the server.
According to [HLS RFC](https://datatracker.ietf.org/doc/html/rfc8216):
* "the client
   SHOULD NOT choose a segment that starts less than three target
   durations from the end of the Playlist file." 
* "If the client reloads a Playlist file and finds that it has not
   changed, then it MUST wait for a period of one-half the target
   duration before retrying."